engine/src/main/scala/cromwell/engine/workflow/lifecycle/MaterializeWorkflowDescriptorActor.scala

```diff
@@ -475,22 +438,56 @@ class MaterializeWorkflowDescriptorActor(serviceRegistryActor: ActorRef,
       list.sequence[Checked, Unit].void
     }
 
+    val baseResolvers: List[String => WorkflowSource] = if (importLocalFilesystem) {
+      List(WdlNamespace.fileResolver, WdlNamespace.httpResolver)
+    } else {
+      List(WdlNamespace.httpResolver)
+    }
+
+    import common.validation.Validation._
+
     val wdlNamespaceValidation = source match {
-      case w: WorkflowSourceFilesWithDependenciesZip => validateNamespaceWithImports(w)
+      case w: WorkflowSourceFilesWithDependenciesZip =>
+        for {
+          importsDir <- validateImportsDirectory(w.importsZip)
+          betterFilesImportsDir = better.files.File(importsDir.pathAsString)
+          directoryResolver = WdlNamespace.directoryResolver(betterFilesImportsDir): String => WorkflowSource
+          resolvers = directoryResolver +: baseResolvers
+          wf <- WdlNamespaceWithWorkflow.load(w.workflowSource, resolvers).toErrorOr
+          _ = importsDir.delete(swallowIOExceptions = true)
+        } yield wf
       case w: WorkflowSourceFilesWithoutImports =>
-        val importResolvers: Seq[ImportResolver] = if (importLocalFilesystem) {
-          List(WdlNamespace.fileResolver, WdlNamespace.httpResolver)
-        } else {
-          List(WdlNamespace.httpResolver)
-        }
-        WdlNamespaceWithWorkflow.load(w.workflowSource, importResolvers) match {
-          case Failure(e) => s"Unable to load namespace from workflow: ${e.getMessage}".invalidNel
-          case Success(namespace) => validateWorkflowNameLengths(namespace)
+        WdlNamespaceWithWorkflow.load(w.workflowSource, baseResolvers).toErrorOr
+    }
+
+    /* Publish `MetadataEvent`s for all imports in this `WdlNamespace`. */
+    def publishImportMetadata(wdlNamespace: WdlNamespace): Unit = {
+      def metadataEventForImportedNamespace(ns: WdlNamespace): MetadataEvent = {
+        import MetadataKey._
+        import WorkflowMetadataKeys._
+        // This should only be called on namespaces that are known to have a defined `importUri` so the .get is safe.
+        val escapedUri = ns.importUri.get.escapeMeta
+        MetadataEvent(MetadataKey(
+          workflowIdForLogging, None, SubmissionSection, SubmissionSection_Imports, escapedUri), MetadataValue(ns.sourceString))
+      }
+
+      // Descend the namespace looking for imports and construct `MetadataEvent`s for them.
+      def collectImportEvents(ns: WdlNamespace): List[MetadataEvent] = {
+        // Do not publish import events for URIs which correspond to literal strings as these are the top-level
+        // submitted workflow.
+        val nonTopLevelEvents = ns.importUri.toList collect {
+          case uri if uri != WdlNamespace.WorkflowResourceString => metadataEventForImportedNamespace(ns)
         }
+        nonTopLevelEvents ++ (ns.namespaces.toList map metadataEventForImportedNamespace)
```

Comment:
Does this cover recursive imports ?

Reply:
An earlier version of this did but I don't think this one does. 😦 Good catch!